### PR TITLE
Refactor/arrebol package

### LINKED ISSUE #96: Refactor arrebol package
The SAPS code is closely linked to Arrebol job execution service, it should make it more flexible for new services to be added.

### PR DESCRIPTION
**Pull request by wesleymonte**
Fix #96

### CODE REVIEW

File: src/main/java/org/fogbowcloud/saps/engine/core/scheduler/executor/JobExecutionService.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package org.fogbowcloud.saps.engine.core.scheduler.executor;
+
+import java.util.List;
+import org.fogbowcloud.saps.engine.core.scheduler.executor.arrebol.dtos.JobResponseDTO;
+import org.fogbowcloud.saps.engine.core.model.SapsJob;
+
+public interface JobExecutionService {
+
+    public String submit(SapsJob job) throws Exception;
```
Comment (thiagoyeds) [inline src/main/java/org/fogbowcloud/saps/engine/core/scheduler/executor/JobExecutionService.java]: Remove public terms from the methods on this interface

File: src/main/java/org/fogbowcloud/saps/engine/core/scheduler/executor/JobExecutionService.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+    public long getWaitingJobs() throws Exception;
+
+    public JobResponseDTO getJob(String jobId) throws Exception;
```
Comment (thiagoyeds) [inline src/main/java/org/fogbowcloud/saps/engine/core/scheduler/executor/JobExecutionService.java]: Rename this method to `getJobById`

File: src/main/java/org/fogbowcloud/saps/engine/core/scheduler/executor/arrebol/dtos/TaskSpecResponseDTO.java [hunk 3]
```diff
@@ -1,8 +1,9 @@
-package org.fogbowcloud.saps.engine.core.dto;
+package org.fogbowcloud.saps.engine.core.scheduler.executor.arrebol.dtos;
 
 import java.io.Serializable;
 import java.util.List;
 import java.util.Map;
+import org.fogbowcloud.saps.engine.core.scheduler.executor.arrebol.dtos.CommandResponseDTO;
```
Comment (thiagoyeds) [inline src/main/java/org/fogbowcloud/saps/engine/core/scheduler/executor/arrebol/dtos/TaskSpecResponseDTO.java]: Remove it

File: src/main/java/org/fogbowcloud/saps/engine/core/scheduler/Scheduler.java [hunk 4]
```diff
@@ -38,23 +43,26 @@
     //FIXME Remove properties field and add new variables
     private Properties properties;
     private Catalog catalog;
-    private Arrebol arrebol;
+    private JobExecutionService jobExecutionService;
+    //FIXME Instance as Thread-Safe object
+    private List<JobSubmitted> submittedJobsID;
```
Comment (thiagoyeds) [inline src/main/java/org/fogbowcloud/saps/engine/core/scheduler/Scheduler.java]: Rename this attribute to `submittedJobs`

File: src/main/java/org/fogbowcloud/saps/engine/core/scheduler/Scheduler.java [hunk 5]
```diff
@@ -38,23 +43,26 @@
     //FIXME Remove properties field and add new variables
     private Properties properties;
     private Catalog catalog;
-    private Arrebol arrebol;
+    private JobExecutionService jobExecutionService;
```
Comment (thiagoyeds) [inline src/main/java/org/fogbowcloud/saps/engine/core/scheduler/Scheduler.java]: Add `final` to vars (line 44 to 46)

File: src/main/java/org/fogbowcloud/saps/engine/core/scheduler/Scheduler.java [hunk 6]
```diff
@@ -38,23 +43,26 @@
...
+    //FIXME Instance as Thread-Safe object
+    private List<JobSubmitted> submittedJobsID;
 
     public Scheduler(Properties properties) throws SapsException {
     	this(properties, new JDBCCatalog(properties), Executors.newScheduledThreadPool(1),
-                new DefaultArrebol(properties), new DefaultRoundRobin());
+                new ArrebolJobExecutionService(new ArrebolRequestsHelper(properties.getProperty(SapsPropertiesConstants.ARREBOL_BASE_URL))), new DefaultRoundRobin());
```
Comment (thiagoyeds) [inline src/main/java/org/fogbowcloud/saps/engine/core/scheduler/Scheduler.java]: Receive the properties in the ArrebolRequestHelper class and create a check to verify that the Arrebol base URL has been set.

File: src/main/java/org/fogbowcloud/saps/engine/core/scheduler/Scheduler.java [hunk 7]
```diff
@@ -85,33 +93,52 @@ public void start() {
...
+                rollBackTaskState(task);
+            else if (jobsWithEqualJobName.size() == 1) { // TODO add check jobName ==
+                // jobsWithEqualJobName.get(0).get...
+                String arrebolJobId = jobsWithEqualJobName.get(0).getId();
+                updateStateInCatalog(task, task.getState(), SapsImage.AVAILABLE, SapsImage.NON_EXISTENT_DATA,
+                    arrebolJobId,
+                    "updates task [" + task.getTaskId() + "] with Arrebol job ID [" + arrebolJobId + "]");
+                this.submittedJobsID.add(new JobSubmitted(task.getArrebolJobId(), task));
+            } else {
+                // TODO ????
```
Comment (thiagoyeds) [inline src/main/java/org/fogbowcloud/saps/engine/core/scheduler/Scheduler.java]: Change the message of this TODO to "discuss what action to take in this situation. Should I remove all jobs found from the execution service and do a rollback in task? Or should I take advantage of any of them?"

File: src/main/java/org/fogbowcloud/saps/engine/core/scheduler/Scheduler.java [hunk 8]
```diff
@@ -130,20 +157,24 @@ private void rollBackTaskState(SapsImage task) {
...
+        }
     }
 
     /**
      * This function selects tasks following a strategy for submit in Arrebol.
      *
      * @return selected tasks list
      */
-    protected List<SapsImage> selectTasks() {
+    List<SapsImage> selectTasks() throws Exception {
```
Comment (thiagoyeds) [inline src/main/java/org/fogbowcloud/saps/engine/core/scheduler/Scheduler.java]: Add some visible encapsulation to this method, `public` in this case.

File: src/main/java/org/fogbowcloud/saps/engine/core/scheduler/Scheduler.java [hunk 9]
```diff
@@ -373,7 +404,7 @@ else if (state == ImageTaskState.PREPROCESSING)
      * submitted jobs to Arrebol.
      */
     private void checker() {
-        List<JobSubmitted> submittedJobs = arrebol.returnAllJobsSubmitted();
+        List<JobSubmitted> submittedJobs = this.submittedJobsID;
```
Comment (thiagoyeds) [inline src/main/java/org/fogbowcloud/saps/engine/core/scheduler/Scheduler.java]: Do not create a new collection, use the `submittedJobs` attribute of this class

File: src/main/java/org/fogbowcloud/saps/engine/core/scheduler/Scheduler.java [hunk 10]
```diff
@@ -383,48 +414,50 @@ private void checker() {
...
                             SapsImage.NONE_ARREBOL_JOB_ID, "updates task [" + task.getTaskId() + "] with failed state");
+                    }
+
+                    addTimestampTaskInCatalog(task, "updates task [" + task.getTaskId() + "] timestamp");
+
+                    finishedJobs.add(job);
+                } else {
+                    LOGGER.info("Job [" + jobId + "] has NOT been finished");
                 }
+            } catch (ConnectException e) {
```
Comment (thiagoyeds) [inline src/main/java/org/fogbowcloud/saps/engine/core/scheduler/Scheduler.java:491]: log it at least

File: src/main/java/org/fogbowcloud/saps/engine/core/scheduler/executor/JobExecutionService.java [hunk 11]
```diff
@@ -0,0 +1,17 @@
+package org.fogbowcloud.saps.engine.core.scheduler.executor;
+
+import java.util.List;
+import org.fogbowcloud.saps.engine.core.scheduler.executor.arrebol.dtos.JobResponseDTO;
+import org.fogbowcloud.saps.engine.core.model.SapsJob;
+
+public interface JobExecutionService {
```
Comment (thiagoyeds) [inline src/main/java/org/fogbowcloud/saps/engine/core/scheduler/executor/JobExecutionService.java:8]: Wouldn't it be better to create an exception for this interface and use it in methods?

**Review by thiagoyeds — CHANGES_REQUESTED** [review]
@wesleymonte this PR brought a very good thing, which is the ease of a new implementation of the job execution service, which is currently Arrebol. However changes were made that resulted in several errors in the SchedulerTest, try to resolve them in addition to the ones I already requested in my review and it looks good to me.

File: src/main/java/org/fogbowcloud/saps/engine/core/scheduler/Scheduler.java (hunk 5 above)
Comment (wesleymonte) [reply src/main/java/org/fogbowcloud/saps/engine/core/scheduler/Scheduler.java]: Done.

File: src/main/java/org/fogbowcloud/saps/engine/core/scheduler/executor/JobExecutionService.java (hunk 1 above)
Comment (wesleymonte) [reply src/main/java/org/fogbowcloud/saps/engine/core/scheduler/executor/JobExecutionService.java]: Done.

File: src/main/java/org/fogbowcloud/saps/engine/core/scheduler/executor/JobExecutionService.java (hunk 2 above)
Comment (wesleymonte) [reply src/main/java/org/fogbowcloud/saps/engine/core/scheduler/executor/JobExecutionService.java]: Done.

File: src/main/java/org/fogbowcloud/saps/engine/core/scheduler/executor/arrebol/dtos/TaskSpecResponseDTO.java (hunk 3 above)
Comment (wesleymonte) [reply src/main/java/org/fogbowcloud/saps/engine/core/scheduler/executor/arrebol/dtos/TaskSpecResponseDTO.java]: Done.

File: src/main/java/org/fogbowcloud/saps/engine/core/scheduler/Scheduler.java (hunk 4 above)
Comment (wesleymonte) [reply src/main/java/org/fogbowcloud/saps/engine/core/scheduler/Scheduler.java]: Done.

File: src/main/java/org/fogbowcloud/saps/engine/core/scheduler/Scheduler.java (hunk 5 above)
Comment (wesleymonte) [reply src/main/java/org/fogbowcloud/saps/engine/core/scheduler/Scheduler.java]: Done.

File: src/main/java/org/fogbowcloud/saps/engine/core/scheduler/Scheduler.java (hunk 6 above)
Comment (wesleymonte) [reply src/main/java/org/fogbowcloud/saps/engine/core/scheduler/Scheduler.java]: The properties now only exist during the construction of the Scheduler. Arrebol Request Helper only needs one field, so there is no point in passing properties to it.

File: src/main/java/org/fogbowcloud/saps/engine/core/scheduler/Scheduler.java (hunk 7 above)
Comment (wesleymonte) [reply src/main/java/org/fogbowcloud/saps/engine/core/scheduler/Scheduler.java]: Done.

File: src/main/java/org/fogbowcloud/saps/engine/core/scheduler/Scheduler.java (hunk 8 above)
Comment (wesleymonte) [reply src/main/java/org/fogbowcloud/saps/engine/core/scheduler/Scheduler.java]: Done.

File: src/main/java/org/fogbowcloud/saps/engine/core/scheduler/Scheduler.java (hunk 9 above)
Comment (wesleymonte) [reply src/main/java/org/fogbowcloud/saps/engine/core/scheduler/Scheduler.java]: Done.

File: src/main/java/org/fogbowcloud/saps/engine/core/scheduler/Scheduler.java (hunk 10 above)
Comment (wesleymonte) [reply src/main/java/org/fogbowcloud/saps/engine/core/scheduler/Scheduler.java:491]: Done.

**Comment by thiagoyeds** [comment]
This repo does not contain code

**Pull request closed without merging**